libs/domain/checkout/src/services/state/default-checkout-state.service.ts

```diff
@@ -0,0 +1,93 @@
+import { StorageService, StorageType } from '@spryker-oryx/core';
+import { inject } from '@spryker-oryx/di';
+import { BehaviorSubject, map, Observable, take, tap } from 'rxjs';
+import { Checkout, checkoutDataStorageKey } from '../../models';
+import { CheckoutStateService } from './checkout-state.service';
+
+type m = { valid?: boolean; value?: Partial<Checkout[keyof Checkout]> | null };
+
+export class DefaultCheckoutStateService implements CheckoutStateService {
+  protected subject = new BehaviorSubject<Map<keyof Checkout, m>>(new Map());
+
+  constructor(protected storage = inject(StorageService)) {
+    this.restore();
+  }
+
+  set<K extends keyof Checkout>(
+    key: K,
+    item: {
+      valid?: boolean;
+      value?: Partial<Checkout[K]> | null;
+    }
+  ): void {
+    const collected = this.subject.value;
+    const existing = collected.get(key);
+    if (existing && item.valid !== undefined) existing.valid = item.valid;
+    if (existing && item.value !== undefined) existing.value = item.value;
+    if (!existing) collected.set(key, item);
+    this.subject.next(collected);
+    this.storage.set(
+      checkoutDataStorageKey,
+      Array.from(collected),
+      StorageType.SESSION
+    );
+  }
+
+  get<K extends keyof Checkout>(key: K): Observable<Checkout[K] | null> {
+    return this.subject.pipe(
+      map((data) => {
+        if (!data.get(key)) this.set(key, {});
+        return data.get(key)?.value as Checkout[K] | null;
+      })
```

Comment:
let's maybe add `distinctUntilChanged` to improve performance (now all subsribers to data will get emission whenever any data will change)